### PR TITLE
test: clean up remaining core function tests, introduce testing utilities and an ISO helper

### LINKED ISSUE #215: core package test refactor
As a contributor to Code PushUp, I would like for the tests to be consistently structured and follow good practices.
There are several areas in which I identified room for improvement. In this issue, the following targeted changes should be made.

## Coverage
- [x] Add a happy path test to `read-code-pushup-config` and invalid config error scenario.
- [x] Test both the regular approach and ESM plugin approach once #248 is merged.

In both `collect` and `collect-and-persist`, the main cases should be tested. This means:
- [x] Add a test case for no plugins passed to `collect`.
- [x] Add a test case for explicitly adding multiple output options for `collect-and-persist` and explicitly mark each use case.

## Files
These tests are mostly in scope of **integration tests** (except for `json-to-gql`). Hence ,they should be categorised as integration tests and structured accordingly.
**Note**: `nx` configuration of integration vs unit tests is OOS. This is only about the location and naming of the files.
- [x] Each test file name reflects that it is an integration test (to be finalised).
- [x] Move all configurations and mocks for FS and other under `testing-utils`.
- [x] Remove the unused `types.ts` file.
- [x] It should be decided whether `read-code-pushup-config` and `json-to-gql` and the progress configs belong to this package - in my opinion, the first two functions belong in `utils` and the configurations should be moved to `testing-utils`. However, perhaps there is a reason to keep them in `core`?
- [x] Keep only one `README` file. Currently, there are two - one in the root folder and one under `docs`.
- [x] Merge `DEFAULT_TESTING_CLI_OPTIONS` from `constant.ts`, also used in other packages, under `testing-utils` or use hardcoded values for better readability where applicable.

## Good practices
In order to create and maintain tests easily, their readability should be prioritised. This means minimising boiler plate code, keeping the test flow linear and naming things based on their purpose.
- [x] Instead of mocking console, use `spyOn` for `console.log` in combination with `shouldHaveBeenCalledWith()`.
- [x] Move `portal-client`, `fs` and `fs/promises` mock setup to `testing-utils` and share setup for all FS vs upload flows there.
- [x] Remove boiler plate regarding `pluginConfig` in `execute-plugin`. Having one valid configuration and then overriding its options without the need for helper functions would increase the readability of the tests.
- [x] Remove branching logic in `collect-and-persist` and explicitly state which formats are tested in which file.
- [x] Improve test descriptions and assertions purpose in `persist` - it is not clear when JSON vs `stdout` vs `MD` is created.
- [x] Fix `not.toThrow('no such file...')` in `persist` to either check thrown error message or to check the result did not throw (and did not produce any message).
- [x] Reduce usage of `reportPath` and replace it with a hardcoded value - all tests expect the same path and concrete file format(s). If we want to test different paths, let's create a separate isolated test for that.
- [x] Improve naming of functions `report()`, `auditReport()` and other to signify they are mocks, create only one version and move that to `testing-utils` to be reused.
- [x] Replace `date` field validation in `execute-plugin` with a custom `toBeISODate()` assertion.
- [x] Mock `fs` in `read-code-pushup-config` tests. While `bundleRequire` should not be mocked, it doesn't matter where the file it parses lives. right now, unnecessary one-line files were added.

Note: Remaining coverage related tasks were moved to a separate issue #356

### PR DESCRIPTION
**Pull request by Tlacenka**
Closes #215 #264 

This PR is the last one for the core package test refactor. In it, I do the following:
- Minimise test scope (this means mocking unrelated logic, moving tests to minimal entity).
- Improve test descriptions and test flow.
- Introduce ISO Regexp to help with checking date values.
- Introduce minimal mocks to be reused when a config is required.
- I fix the coverage CI (previously, core tests were failing) by adding `integration-test` target to `testing-utils` and minimising fs dependence of core tests. Run [here](https://github.com/code-pushup/cli/actions/runs/7116779120). 🚥 🟢 
- `testing-utils` are now buildable due to nx buildable rule.

### CODE REVIEW

**Review by MishaSeredenkoPushBased — DISMISSED** [review]
LGTM 👍

**Review by matejchalk — COMMENTED** [review]
Great now all our tests finally pass when run at once :tada: Big step forward for stability :clap:

File: testing-utils/project.json [hunk 1]
```diff
@@ -18,6 +29,14 @@
         "config": "testing-utils/vite.config.unit.ts",
         "reportsDirectory": "../coverage/testing-utils/unit-tests"
       }
+    },
+    "integration-test": {
```
[the remark this answers is not in the document]
Comment (Tlacenka) [reply testing-utils/project.json:33]: Good point.

File: packages/core/src/lib/upload.unit.test.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        server: 'https://example.com/api',
+        organization: 'code-pushup',
+        project: 'cli',
+      },
+      persist: {
+        outputDir: MEMFS_VOLUME,
+        filename: 'report',
+        format: ['json'],
+      },
+    } as UploadOptions);
```
[the remark this answers is not in the document]
Comment (Tlacenka) [reply packages/core/src/lib/upload.unit.test.ts:34]: Good point, removed.

File: packages/core/src/lib/collect-and-persist.unit.test.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        duration: 0,
+        categories: [],
+        plugins: [],
+      },
+      {
+        ...MINIMAL_CONFIG_MOCK,
+        verbose: false,
+        progress: false,
+      },
+    );
```
[the remark this answers is not in the document]
Comment (Tlacenka) [reply packages/core/src/lib/collect-and-persist.unit.test.ts:62]: I suppose you are right.

File: packages/core/src/lib/collect-and-persist.unit.test.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+          packageName: 'code-pushup',
+          version: '0.0.1',
+          date: new Date().toISOString(),
+          duration: 0,
+          categories: [],
+          plugins: [],
+        } as Report),
+    ),
+  };
+});
```
[the remark this answers is not in the document]
Comment (Tlacenka) [reply packages/core/src/lib/collect-and-persist.unit.test.ts:25]: Ah, fair point. I thought I had to use it in order to check `shouldHaveBeenCalled`.

File: packages/core/src/lib/collect-and-persist.unit.test.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+vi.mock('./implementation/persist', async () => {
+  return {
+    persistReport: vi.fn().mockImplementation(async () => ({})),
+    logPersistedResults: vi.fn().mockReturnValue(undefined),
+  };
+});
```
[the remark this answers is not in the document]
Comment (Tlacenka) [reply packages/core/src/lib/collect-and-persist.unit.test.ts:32]: I suppose it doesn't as I only check the function was called.

File: testing-utils/src/lib/fixtures/configs/minimal-config.mock.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  args: ['-v'],
+  outputFile: 'output.json',
+};
+
+export const MINIMAL_RUNNER_FUNCTION_MOCK: RunnerFunction = () => [
+  {
+    slug: 'node-version',
+    title: 'Node version',
+    description: 'Returns node version',
+    docsUrl: 'https://nodejs.org/',
```
[the remark this answers is not in the document]
Comment (Tlacenka) [reply testing-utils/src/lib/fixtures/configs/minimal-config.mock.ts:22]: I see, thanks for pointing it out.

File: testing-utils/package.json
[the remark this answers is not in the document]
Comment (Tlacenka) [reply testing-utils/package.json:1]: Done!

File: packages/core/src/lib/collect-and-persist.unit.test.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    plugins: [],
+  } as Report),
+}));
+
+vi.mock('./implementation/persist', async () => {
+  return {
+    persistReport: vi.fn(),
+    logPersistedResults: vi.fn(),
+  };
+});
```
Comment (matejchalk) [inline packages/core/src/lib/collect-and-persist.unit.test.ts:27]:
```suggestion
vi.mock('./implementation/persist', () => ({
  persistReport: vi.fn(),
  logPersistedResults: vi.fn(),
}));
```

**Review by matejchalk — APPROVED** [review]

File: packages/core/src/lib/collect-and-persist.unit.test.ts (hunk 7 above)
Comment (Tlacenka) [reply packages/core/src/lib/collect-and-persist.unit.test.ts]: Oops, fixed.

**Review by matejchalk — APPROVED** [review]

**Pull request merged**